File: jobhunt/matching.py
"""Match scoring and query building.

Emulates the original artex_job_hunt scoring: each listing gets a
deterministic score out of 50, rendered as ``[████░░░░░░] 38% (19/50)``.

Components (max 50):
- title overlap with the user's resume-derived titles ........ up to 20
- skills found in the listing title/description/tags ......... up to 15
- duty keywords found in the description ..................... up to 10
- certifications mentioned ................................... up to 5
"""
from __future__ import annotations

import re
from dataclasses import dataclass

from . import config
from .scrapers.base import JobPosting, SearchQuery

STOPWORDS = {
    "of", "and", "the", "a", "an", "for", "to", "in", "at", "on", "with",
    "or", "&", "-", "i", "ii", "iii", "sr", "jr", "senior", "junior",
}


def _tokens(text: str) -> set[str]:
    return {
        t
        for t in re.findall(r"[a-z0-9+#/.]+", text.lower())
        if t not in STOPWORDS and len(t) > 1
    }


@dataclass
class ScoreBreakdown:
    title: int = 0
    skills: int = 0
    duties: int = 0
    certifications: int = 0

    @property
    def total(self) -> int:
        return min(
            config.MAX_SCORE,
            self.title + self.skills + self.duties + self.certifications,
        )


def score_posting(posting: JobPosting, profile: dict) -> ScoreBreakdown:
    """Score one posting against the merged resume profile."""
    breakdown = ScoreBreakdown()
    title_tokens = _tokens(posting.title)
    haystack = " ".join(
        [posting.title, posting.description, " ".join(posting.tags)]
    ).lower()

    # Title overlap: best overlap ratio against any profile title.
    best_ratio = 0.0
    for profile_title in profile.get("titles", []):
        pt = _tokens(profile_title)
        if not pt:
            continue
        overlap = len(pt & title_tokens) / len(pt)
        best_ratio = max(best_ratio, overlap)
    breakdown.title = round(best_ratio * 20)

    # Skills: 3 points each, capped at 15.
    hits = 0
    for skill in profile.get("skills", []):
        pattern = r"(?<![\w+#])" + re.escape(skill.lower()) + r"(?![\w+#])"
        if re.search(pattern, haystack):
            hits += 1
    breakdown.skills = min(15, hits * 3)

    # Duties: shared meaningful words between duty bullets and description.
    if posting.description:
        duty_tokens: set[str] = set()
        for duty in profile.get("duties", []):
            duty_tokens |= _tokens(duty)
        desc_tokens = _tokens(posting.description)
        shared = len(duty_tokens & desc_tokens)
        breakdown.duties = min(10, shared // 3)

    # Certifications: 5 if any cert is mentioned.
    for cert in profile.get("certifications", []):
        if cert.lower() in haystack:
            breakdown.certifications = 5
            break

    return breakdown


def match_bar(score: int, max_score: int = config.MAX_SCORE) -> str:
    """Render the artex-style match bar, e.g. '[████░░░░░░] 38% (19/50)'."""
    pct = round(score / max_score * 100)
    filled = round(score / max_score * 10)
    return f"[{'█' * filled}{'░' * (10 - filled)}] {pct}% ({score}/{max_score})"


def location_in_range(posting_location: str, preferred: list[str]) -> bool:
    """True if the posting's location matches any preferred location.

    Matching is loose: a preferred location matches if its city token
    appears in the posting location, and 'Remote' matches the usual
    remote/anywhere/flexible phrasings. No preferences = everything passes.
    """
    if not preferred:
        return True
    loc = posting_location.lower()
    for pref in preferred:
        p = pref.lower().strip()
        if not p:
            continue
        if p in {"remote", "anywhere"}:
            if any(
                w in loc for w in ("remote", "anywhere", "flexible", "worldwide")
            ) or loc == "":
                return True
            continue
        city = p.split(",")[0].strip()
        if city and city in loc:
            return True
    return False


def build_queries(profile: dict, locations: list[str]) -> list[SearchQuery]:
    """Fan out (title x location) queries, like the artex cron did."""
    titles = profile.get("titles", [])[:8]
    if not titles:
        # Fall back to top skills as keywords so new users still get results.
        titles = profile.get("skills", [])[:4]
    locs = locations or ["Remote"]
    queries = [
        SearchQuery(keywords=title, location=loc)
        for title in titles
        for loc in locs
    ]
    return queries[: config.MAX_QUERIES_PER_SOURCE]

File: android/core/src/main/kotlin/com/jobhunt/core/Matching.kt
package com.jobhunt.core

import kotlin.math.min
import kotlin.math.roundToInt

/**
 * Match scoring and query building.
 *
 * Each listing gets a deterministic score out of [MAX_SCORE], rendered as
 * `[████░░░░░░] 38% (19/50)`.
 *
 *  - title overlap with resume-derived titles ....... up to 20
 *  - skills found in title/description/tags ......... up to 15
 *  - duty keywords found in the description ......... up to 10
 *  - certifications mentioned ....................... up to 5
 */
object Matching {

    private val STOPWORDS = setOf(
        "of", "and", "the", "a", "an", "for", "to", "in", "at", "on", "with",
        "or", "&", "-", "i", "ii", "iii", "sr", "jr", "senior", "junior",
    )

    private val TOKEN_RE = Regex("""[a-z0-9+#/.]+""")

    data class ScoreBreakdown(
        val title: Int = 0,
        val skills: Int = 0,
        val duties: Int = 0,
        val certifications: Int = 0,
    ) {
        val total: Int get() = min(MAX_SCORE, title + skills + duties + certifications)
    }

    internal fun tokens(text: String): Set<String> =
        TOKEN_RE.findAll(text.lowercase())
            .map { it.value }
            .filter { it !in STOPWORDS && it.length > 1 }
            .toSet()

    fun score(posting: JobPosting, profile: SearchProfile): ScoreBreakdown {
        val titleTokens = tokens(posting.title)
        val haystack = buildString {
            append(posting.title).append(' ')
            append(posting.description).append(' ')
            append(posting.tags.joinToString(" "))
        }.lowercase()

        // Title: best token-overlap ratio against any profile title.
        var bestRatio = 0.0
        for (profileTitle in profile.titles) {
            val pt = tokens(profileTitle)
            if (pt.isEmpty()) continue
            bestRatio = maxOf(bestRatio, pt.intersect(titleTokens).size.toDouble() / pt.size)
        }
        val titleScore = (bestRatio * 20).roundToInt()

        // Skills: 3 points each, capped at 15.
        val skillHits = profile.skills.count { skill ->
            Regex("""(?<![\w+#])${Regex.escape(skill.lowercase())}(?![\w+#])""")
                .containsMatchIn(haystack)
        }
        val skillScore = min(15, skillHits * 3)

        // Duties: shared meaningful words between duty bullets and description.
        val dutyScore = if (posting.description.isNotBlank()) {
            val dutyTokens = profile.duties.flatMap { tokens(it) }.toSet()
            min(10, dutyTokens.intersect(tokens(posting.description)).size / 3)
        } else 0

        val certScore = if (profile.certifications.any { it.lowercase() in haystack }) 5 else 0

        return ScoreBreakdown(titleScore, skillScore, dutyScore, certScore)
    }

    /** Renders the match bar, e.g. `[████░░░░░░] 38% (19/50)`. */
    fun matchBar(score: Int, maxScore: Int = MAX_SCORE): String {
        val pct = (score * 100.0 / maxScore).roundToInt()
        val filled = (score * 10.0 / maxScore).roundToInt().coerceIn(0, 10)
        return "[" + "█".repeat(filled) + "░".repeat(10 - filled) + "] $pct% ($score/$maxScore)"
    }

    /**
     * True if a posting's location matches any preferred location. Matching is
     * loose: a preference matches when its city token appears in the posting
     * location, and "Remote" matches the usual remote/anywhere phrasings.
     * No preferences means everything passes.
     */
    fun locationInRange(postingLocation: String, preferred: List<String>): Boolean {
        if (preferred.isEmpty()) return true
        val loc = postingLocation.lowercase()
        for (pref in preferred) {
            val p = pref.lowercase().trim()
            if (p.isEmpty()) continue
            if (p == "remote" || p == "anywhere") {
                val remoteish = listOf("remote", "anywhere", "flexible", "worldwide")
                if (loc.isEmpty() || remoteish.any { it in loc }) return true
                continue
            }
            val city = p.substringBefore(',').trim()
            if (city.isNotEmpty() && city in loc) return true
        }
        return false
    }

    /** Fan out (title x location) queries, capped at [MAX_QUERIES_PER_SOURCE]. */
    fun buildQueries(profile: SearchProfile, locations: List<String>): List<SearchQuery> {
        val titles = profile.titles.take(8).ifEmpty { profile.skills.take(4) }
        val locs = locations.ifEmpty { listOf("Remote") }
        return titles
            .flatMap { title -> locs.map { loc -> SearchQuery(title, loc) } }
            .take(MAX_QUERIES_PER_SOURCE)
    }
}
